### PR TITLE
feat(*): [RO-18670] add ability to ignore dependencies

### PR DESCRIPTION
**Pull request by terry-au**
# Changes

- Adds the ability to mark a dependency as ignored by setting `spurIocIgnore` to `true` on the dependency.
- Update warning output when dependencies are being overridden.

### CODE REVIEW

File: src/ContainerManagement.js [hunk 1]
```diff
@@ -1,35 +1,60 @@
...
-      this.warnIfNeeded(name);
+      this.warnIfNeeded(name, dependency);
     }
     this.dependencies[name] = Dependency.resolvableDependency(name, dependency);
     return this;
   },
 
   addDependency(name, dependency, suppressWarning = false) {
+    if (this.shouldIgnoreDependency(dependency)) {
+      return this;
```
Comment (acolchado) [inline src/ContainerManagement.js:54]: Wondering if we should add a debug message here that says we are ignoring this dependency..

File: src/ContainerManagement.js [hunk 2]
```diff
@@ -1,35 +1,60 @@
...
+  getDependencySourceHint(dependency) {
+    if (_isFunction(dependency)) {
+      return _get(dependency, 'name') || '<anonymous function>';
+    }
+
+    if (_isObject(dependency)) {
+      return _get(dependency, 'constructor.name') || '<object>';
+    }
+
+    return '<primitive>';
```
Comment (acolchado) [inline src/ContainerManagement.js]: Can we print the type? May help with the debugging.

**Review by acolchado — CHANGES_REQUESTED** [review]
Overall this looks like it will provide us with the filtering we need. A couple of minor things:

`Can you add coverage for all of the public (documented) [registration entry points](ioc.registerDependencies)?`

- ~ioc.registerDependencies~
- ioc.registerFolders
- ioc.addResolvableDependency
- ioc.addDependency

`Could we filter out *.d.ts files?`

I made it too relaxed with the 'ts' inclusion. But this doesn't have to be in this PR and I can also take care of it.

File: src/ContainerManagement.js [hunk 3]
```diff
@@ -1,35 +1,60 @@
 const _forEach = require('lodash.foreach');
+const _get = require('lodash.get');
```
Comment (acolchado) [inline src/ContainerManagement.js:2]: Doesn't seem to be registered as a dependency of this package.

File: src/ContainerManagement.js (hunk 1 above)
Comment (terry-au) [reply src/ContainerManagement.js:54]: @acolchado updated to print a warning if `suppressWarning` is false.

File: src/ContainerManagement.js (hunk 2 above)
Comment (terry-au) [reply src/ContainerManagement.js]: Updated 👍

**Review by acolchado — APPROVED** [review]

**Pull request merged**